sdk/tests/deqp/framework/referencerenderer/rrRenderer.js

```diff
@@ -550,6 +522,372 @@ rrRenderer.clipTest = function(x, y, z, rect) {
+        var sy = [sy0, sy0, sy1, sy1];
+
+        // Viewport test
+        var outX1 = x0 + 1 == this.m_viewport.rect.left + this.m_viewport.rect.width;
+        var outY1 = y0 + 1 == this.m_viewport.rect.bottom + this.m_viewport.rect.height;
+
+        // Coverage
+        var coverage = 0;
+
+        // Evaluate edge values
+        var e01 = [];
+        var e12 = [];
+        var e20 = [];
+        for (var i = 0; i < 4; i++) {
+            e01.push(this.m_edge01.evaluateEdge(sx[i], sy[i]));
+            e12.push(this.m_edge12.evaluateEdge(sx[i], sy[i]));
+            e20.push(this.m_edge20.evaluateEdge(sx[i], sy[i]));
+        }
+
+        // Compute coverage mask
+        coverage = rrRenderer.setCoverageValue(coverage, 1, 0, 0, 0, this.m_edge01.isInsideCCW(e01[0]) && this.m_edge12.isInsideCCW(e12[0]) && this.m_edge20.isInsideCCW(e20[0]));
+        coverage = rrRenderer.setCoverageValue(coverage, 1, 1, 0, 0, !outX1 && this.m_edge01.isInsideCCW(e01[1]) && this.m_edge12.isInsideCCW(e12[1]) && this.m_edge20.isInsideCCW(e20[1]));
+        coverage = rrRenderer.setCoverageValue(coverage, 1, 0, 1, 0, !outY1 && this.m_edge01.isInsideCCW(e01[2]) && this.m_edge12.isInsideCCW(e12[2]) && this.m_edge20.isInsideCCW(e20[2]));
+        coverage = rrRenderer.setCoverageValue(coverage, 1, 1, 1, 0, !outX1 && !outY1 && this.m_edge01.isInsideCCW(e01[3]) && this.m_edge12.isInsideCCW(e12[3]) && this.m_edge20.isInsideCCW(e20[3]));
+
+        // Advance to next location
+        this.m_curPos[0] += 2;
+        if (this.m_curPos[0] > this.m_bboxMax[0]) {
+            this.m_curPos[0] = this.m_bboxMin[0];
+            this.m_curPos[1] += 2;
+        }
+
+        if (coverage == 0)
+            continue; // Discard
+
+        // Compute depth and barycentric coordinates
+        var edgeSum = deMath.add(deMath.add(e01, e12), e20);
+        var z0 = deMath.divide(e12, edgeSum);
+        var z1 = deMath.divide(e20, edgeSum);
+
+        var b0 = deMath.multiply(e12, [this.m_v0[3], this.m_v0[3], this.m_v0[3], this.m_v0[3]]);
+        var b1 = deMath.multiply(e20, [this.m_v1[3], this.m_v1[3], this.m_v1[3], this.m_v1[3]]);
+        var b2 = deMath.multiply(e01, [this.m_v2[3], this.m_v2[3], this.m_v2[3], this.m_v2[3]]);
+        var bSum = deMath.add(deMath.add(b0, b1), b2);
+        var barycentric0 = deMath.divide(b0, bSum);
+        var barycentric1 = deMath.divide(b1, bSum);
+        var barycentric2 = deMath.subtract(deMath.subtract([1, 1, 1, 1], barycentric0), barycentric1);
+
+        // Determine if (x0, y0), (x0 + 1, y0), (x0, y0 + 1), (x0 + 1, y0 + 1) can be rendered
+        for (var fragNdx = 0; fragNdx < 4; fragNdx++) {
+            var xo = fragNdx % 2;
+            var yo = Math.trunc(fragNdx / 2);
+            var x = x0 + xo;
+            var y = y0 + yo;
+
+            // The value of numSamples always equals 1 in sglrReferenceContext
+            if(rrRenderer.getCoverageAnyFragmentSampleLive(coverage, 1, xo, yo)) {
```

Comment:
The value of numSamples always equals 1 in sglrReferenceContext:
https://github.com/KhronosGroup/WebGL/blob/master/sdk/tests/deqp/framework/opengl/simplereference/sglrReferenceContext.js#L3528